drivers/pwm/pwm_numaker.c

```diff
@@ -0,0 +1,574 @@
+	}
+}
+
+static void pwm_numaker_p0_isr(const struct device *dev)
+{
+	/* Pair0 service channel 0, 1 */
+	__pwm_numaker_isr(dev, 0, 1);
+}
+
+static void pwm_numaker_p1_isr(const struct device *dev)
+{
+	/* Pair1 service channel 2, 3 */
+	__pwm_numaker_isr(dev, 2, 3);
+}
+
+static void pwm_numaker_p2_isr(const struct device *dev)
+{
+	/* Pair2 service channel 4, 5 */
+	__pwm_numaker_isr(dev, 4, 5);
+}
+#endif /* CONFIG_PWM_CAPTURE */
+
+/* PWM driver registration */
+static const struct pwm_driver_api pwm_numaker_driver_api = {
+	.set_cycles = pwm_numaker_set_cycles,
+	.get_cycles_per_sec = pwm_numaker_get_cycles_per_sec,
+#ifdef CONFIG_PWM_CAPTURE
+	.configure_capture = pwm_numaker_configure_capture,
+	.enable_capture = pwm_numaker_enable_capture,
+	.disable_capture = pwm_numaker_disable_capture,
+#endif /* CONFIG_PWM_CAPTURE */
+};
+
+/* Alternative EPWM clock get rate before support standard clock_control_get_rate */
+static int __pwm_numaker_clk_get_rate(EPWM_T *epwm, uint32_t *rate)
+{
+	uint32_t clk_src;
+	uint32_t epwm_clk_src;
+
+	if (epwm == EPWM0) {
+		clk_src = CLK->CLKSEL2 & CLK_CLKSEL2_EPWM0SEL_Msk;
+	} else { /* (epwm == EPWM1) */
+		clk_src = CLK->CLKSEL2 & CLK_CLKSEL2_EPWM1SEL_Msk;
+	}
+
+	if (clk_src == 0U) {
+		/* clock source is from PLL clock */
+		epwm_clk_src = CLK_GetPLLClockFreq();
+	} else {
+		/* clock source is from PCLK */
+		SystemCoreClockUpdate();
+		if (epwm == EPWM0) {
+			epwm_clk_src = CLK_GetPCLK0Freq();
+		} else { /* (epwm == EPWM1) */
+			epwm_clk_src = CLK_GetPCLK1Freq();
+		}
+	}
+	*rate = epwm_clk_src;
+	return 0;
+}
+
+static int pwm_numaker_init(const struct device *dev)
+{
+	const struct pwm_numaker_config *cfg = dev->config;
+	struct pwm_numaker_data *data = dev->data;
+	EPWM_T *epwm = cfg->epwm;
+	uint32_t clock_freq;
+	int err = 0;
+
+	struct numaker_scc_subsys scc_subsys;
+
+	SYS_UnlockReg();
+
+	/* CLK controller */
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx = cfg->clk_modidx;
+	scc_subsys.pcc.clk_src = cfg->clk_src;
+	scc_subsys.pcc.clk_div = cfg->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock() */
+	err = clock_control_on(cfg->clk_dev, (clock_control_subsys_t)&scc_subsys);
+	if (err != 0) {
+		goto done;
+	}
+	/* Equivalent to CLK_SetModuleClock() */
+	err = clock_control_configure(cfg->clk_dev, (clock_control_subsys_t)&scc_subsys, NULL);
+	if (err != 0) {
+		goto done;
+	}
+
+	/* Not support standard clock_control_get_rate yet */
+	/* clock_control_get_rate(cfg->clk_dev,(clock_control_subsys_t)&scc_subsys,&clock_freq); */
+	err = __pwm_numaker_clk_get_rate(epwm, &clock_freq);
+
+	if (err < 0) {
+		LOG_ERR("Get EPWM clock rate failure %d", err);
+		goto done;
+	}
+	data->clock_freq = clock_freq;
+	data->cycles_per_sec = data->clock_freq / (cfg->prescale + 1U);
+
+	err = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err) {
+		LOG_ERR("Failed to apply pinctrl state");
+		goto done;
+	}
+
```

Comment:
drop one empty line

Reply:
Thanks and finished all of the above mentioned points at commit  6116dc068b.